src/lib.rs

```diff
@@ -401,22 +427,186 @@ impl Cascade {
 
 impl fmt::Display for Cascade {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(
+        writeln!(
             f,
-            "salt={:?} inverted={} hash_algorithm={}\n",
+            "salt={:?} inverted={} hash_algorithm={}",
             self.salt, self.inverted, self.hash_algorithm,
         )?;
         for filter in &self.filters {
-            write!(f, "\t[{}]\n", filter)?;
+            writeln!(f, "\t[{}]", filter)?;
+        }
+        Ok(())
+    }
+}
+
+enum BuildStatus {
+    Waiting(usize, usize),
+    Finalized,
+}
+
+pub struct CascadeBuilder {
+    filters: Option<Vec<Bloom>>,
+    salt: Option<Vec<u8>>,
+    hash_algorithm: HashAlgorithm,
+    to_include: Vec<CascadeIndexGenerator>,
+    to_exclude: Vec<CascadeIndexGenerator>,
+    status: BuildStatus,
+}
+
+fn new_crlite_bloom(include_capacity: usize, exclude_capacity: usize, top_layer: bool) -> Bloom {
+    assert!(include_capacity != 0 && exclude_capacity != 0);
+
+    let r = include_capacity as f64;
+    let s = exclude_capacity as f64;
+    // Section III.C of CRLite paper
+    //
+    // The desired false positive rate for the top layer is
+    //   p = r/(sqrt(2)*s).
+    // At subsequent layers it is p = 1/2.
+    let log2_fp_rate = match top_layer {
+        true => (r / s).log2() - 0.5f64,
+        false => -1f64,
+    };
+
+    // k = log2(1/p)
+    let n_hash_funcs = (-log2_fp_rate).round() as u32;
+
+    // m = r log2(1/p) / ln(2)
+    let size = max(
+        CASCADE_MIN_LAYER_BITS,
+        (r * (-log2_fp_rate) / (f64::ln(2f64))).round() as u32,
+    );
+
+    Bloom {
+        n_hash_funcs,
+        size,
+        data: vec![0u8; ((size + 7) / 8) as usize],
+    }
+}
+
+impl CascadeBuilder {
+    pub fn new(
+        hash_algorithm: HashAlgorithm,
+        salt: Vec<u8>,
+        include_capacity: usize,
+        exclude_capacity: usize,
+    ) -> Self {
+        CascadeBuilder {
+            filters: Some(vec![new_crlite_bloom(
+                include_capacity,
+                exclude_capacity,
+                true,
+            )]),
+            salt: Some(salt),
+            to_include: vec![],
+            to_exclude: vec![],
+            hash_algorithm,
+            status: BuildStatus::Waiting(include_capacity, exclude_capacity),
+        }
+    }
+
+    pub fn include(&mut self, item: Vec<u8>) {
+        match self.status {
+            BuildStatus::Waiting(ref mut cap, _) if *cap > 0 => *cap -= 1,
+            _ => panic!("capacity violation"),
+        }
+        let mut generator = CascadeIndexGenerator::new(self.hash_algorithm, item);
```

Comment:
It's actually only one copy per call to `Cascade::has` in the runtime code. But maybe `Cascade::has` should just take a `Vec<u8>` instead of `&[u8]`. That will remove a copy in the case where the caller queries a temporary value (which we do in `cert_storage`)

Reply:
Sounds good!